File: model/Formula.cs
using System;
using Newtonsoft.Json;

namespace Com.GitHub.ZachDeibert.CodeTanks.Model {
    public sealed class Formula {
        public enum Type {
            [JsonProperty("polynomial")]
            Polynomial,
            [JsonProperty("exponential")]
            Exponential
        }

        [JsonProperty("min")]
        public double? Minimum;
        [JsonProperty("max")]
        public double? Maximum;
        [JsonProperty("formula")]
        public Type EquationType;
        [JsonProperty("coefficients")]
        public double[] Coefficients;
        [JsonProperty("coefficient")]
        public Formula Coefficient;
        [JsonProperty("base")]
        public Formula Base;
        [JsonProperty("exponent")]
        public Formula Exponent;

        public double Calculate(double x) {
            if ((Minimum != null && x < Minimum) || (Maximum != null && x > Maximum)) {
                throw new ArgumentOutOfRangeException(nameof(x));
            }
            switch (EquationType) {
                case Type.Polynomial:
                    double value = 0;
                    double xPow = 1;
                    for (int i = Coefficients.Length; i >= 0; --i) {
                        value += Coefficients[i] * xPow;
                        xPow *= x;
                    }
                    return value;
                case Type.Exponential:
                    double coefficient = Coefficient?.Calculate(x) ?? 1;
                    double _base = Base?.Calculate(x) ?? 1;
                    double exponent = Exponent?.Calculate(x) ?? 1;
                    return coefficient * Math.Pow(_base, exponent);
                default:
                    throw new NotSupportedException("Invalid equation type");
            }
        }
    }
}

File: apis/java/src/main/java/com/github/zachdeibert/codetanks/api/model/Formula.java
package com.github.zachdeibert.codetanks.api.model;

public final class Formula {
	public enum Type {
		polynomial, exponential
	}

	public Double minimum;
	public Double maximum;
	public Type formula;
	public double[] coefficients;
	public Formula coefficient;
	public Formula base;
	public Formula exponent;

	public double calculate(double x) {
		if ((minimum != null && x < minimum) || (maximum != null && x > maximum)) {
			throw new IllegalArgumentException("x");
		}
		switch (formula) {
			case polynomial:
				double value = 0;
				double xPow = 1;
				for (int i = coefficients.length; i >= 0; --i) {
					value += coefficients[i] * xPow;
					xPow *= x;
				}
				return value;
			case exponential:
				double coefficient = this.coefficient == null ? 1 : this.coefficient.calculate(x);
				double base = this.base == null ? 1 : this.base.calculate(x);
				double exponent = this.exponent == null ? 1 : this.exponent.calculate(x);
				return coefficient * Math.pow(base, exponent);
			default:
				throw new IllegalStateException("Invalid equation type");
		}
	}
}
